Fix testcases on experimental/cas/main

cachemeifyoucan opened the pull request.

cachemeifyoucan committed:
[Lexer] Fix Preprocessor::HandlePragmaMark

Remove duplicated calls due to merge conflict

cachemeifyoucan commented:
@swift-ci test

dexonsmith commented on clang/lib/Lex/Pragma.cpp:
```diff
@@ -421,7 +421,6 @@ void Preprocessor::HandlePragmaMark(Token &MarkTok) {
   SmallString<64> Buffer;
   CurLexer->ReadToEndOfLine(&Buffer);
   if (CurLexer){
-    CurLexer->ReadToEndOfLine(&Buffer);
```
I think you removed the wrong one — you should keep the one inside `if (CurLexer)` not the one that precedes it I think.

cachemeifyoucan commented:
@swift-ci test

dexonsmith commented:
LGTM, even if there are still other tests failing.

cachemeifyoucan committed:
Disable module modification time check when time == 0

CAS file system returns 0 as modification time. Ignore that value for
now.

dexonsmith commented on clang/lib/Frontend/Rewrite/FrontendActions.cpp:
```diff
@@ -185,7 +185,7 @@ RewriteObjCAction::CreateASTConsumer(CompilerInstance &CI, StringRef InFile) {
   CompilerInstance &CI = getCompilerInstance();
   std::unique_ptr<raw_ostream> OS =
-      CI.createDefaultOutputFile(/*Binary=*/true, getCurrentFileOrBufferName());
+      CI.createDefaultOutputFile(/*Binary=*/false, getCurrentFileOrBufferName());
```
I don't understand why this would need to be changed. I don't think the output backend changes anything here.

dexonsmith replied:
(i.e., if it was previously "binary", then it should basically hit the same code paths in OnDiskOutputBackend as it used to hit in CompilerInstance — any fixes should either be in the call to `createFile` or in the outputbackend itself)

dexonsmith commented on clang/lib/Frontend/CompilerInstance.cpp:
```diff
@@ -822,6 +822,7 @@ CompilerInstance::createOutputFileImpl(StringRef OutputPath, bool Binary,
   Expected<std::unique_ptr<OutputFile>> O =
       getOrCreateOutputBackend().createFile(
           OutputPath, OutputConfig()
+                          .setText(!Binary)
```
Can you add a FIXME? Seems like `setTextWithCRLF` should be sufficient.

cachemeifyoucan replied:
Or maybe we define output config to have separate `Text` and `CRLF` field. `.setTextWithCRLF` can be a convenience method in `OutputConfig` to set both.

dexonsmith replied:
Yes, that's what I did over at https://github.com/dexonsmith/llvm-project/commit/21c24f6866048a2f580ffd5cc633c6eed8861862 — a FIXME for now would probably be enough though.

dexonsmith replied:
Updated to also have `setBinary()` and `getBinary()` -- https://github.com/dexonsmith/llvm-project/commit/f037dc0544b6efaff70c206e922a1db4978742bd#diff-0598fa12397ba618b927970836f454d312e68b633ebb634484e7770f0ab9177bR45 -- will send for review soon

dexonsmith commented on clang/lib/Frontend/Rewrite/FrontendActions.cpp:
```diff
@@ -270,7 +270,7 @@ class RewriteIncludesAction::RewriteImportsListener : public ASTReaderListener {
   if (!OutputStream) {
     OutputStream =
-        CI.createDefaultOutputFile(/*Binary=*/true, getCurrentFileOrBufferName());
+        CI.createDefaultOutputFile(/*Binary=*/false, getCurrentFileOrBufferName());
```
Same here: I don't think we should have to touch this code.

cachemeifyoucan replied:
This changes if the raw_fd_stream in OutputFile gets the Text attribute, which will change how `OnDiskOutputFile::takeOSImpl()` works.

Essentially, this will cause the output cannot be piped from stdout. There are tests in clang that uses this behavior so this is a regression.

dexonsmith replied:
What I'm saying is that that takeOSImpl *should* be equivalent to the old logic in CompilerInstance.cpp. If there's a regression, we should fix takeOSImpl to have equivalent logic.

dexonsmith replied:
That case should be hitting this logic:
```
    // Write out the content and close the stream.
    assert(BufferedContent && "Need to write content to a stream");
    OS->write(BufferedContent->begin(), BufferedContent->size());
    OS = nullptr;
```
which should be writing to the pipe.
- Is that logic not getting hit?
- Is `BufferedContent` missing the data?
- Is `OS->write()` not working correctly?

dexonsmith replied:
(this logic should be equivalent to using a `buffer_unique_ostream`, which is what CompilerInstance.cpp had, it's just going through a bunch of other logic on the way there)

cachemeifyoucan replied:
Let me know if the current implementation works better. I teach OnDiskOutputFile to use `buffer_unique_ostream` instead of using `raw_svector_ostream`

cachemeifyoucan replied:
I guess `OnDiskOutputFile::closeFile` is not even hit. I am not sure what you were planned to call it?

dexonsmith replied:
Looks to me like `OnDiskOutputFile::closeFile` is called by both `OnDiskOutputFile::storeContentBuffer()` and `::storeStreamedContent()`, and `OutputFile::close()` should call at least one of them. Is `OutputFile::close()` not getting called?

cachemeifyoucan committed:
[clang] Setup OutputBackend config correctly

Set O_Text for none binary output so the OS setup correctly and the PIPE
output will work with OutputBackend.

dexonsmith commented on llvm/lib/Support/OutputBackend.cpp:
```diff
@@ -583,8 +583,8 @@ std::unique_ptr<raw_pwrite_stream> OnDiskOutputFile::takeOSImpl() {
 
-  // Fall back on the content buffer.
-  return createStreamForContentBuffer();
+  // Fall back to buffer_unique_ostream.
+  return std::make_unique<buffer_unique_ostream>(std::move(OS));
```
I don't think this fixes the underlying issue. Using `raw_svector_stream` (and then dumping to `OS` at the end) should be equivalent to this... since that's exactly what `buffer_unique_ostream` does. But by using `buffer_unique_ostream` directly you're skipping the `ContentBuffer` logic, which is used to mirror between multiple output backends. (Apparently that's not well-enough tested to trigger a test failure in this case...)

Do you have answers to my questions from https://github.com/apple/llvm-project/pull/3285#discussion_r715318211? Note that the new output-backend implementation I'm preparing for review upstream works pretty differently from this, but I think we need to understand the fundamental problem here rather than disabling the code... when I add mirroring behaviour to it I'll need something spiritually to the code you're disabling.

(Note that the other commits look fine. I'm not sure why you've bundled them into a single PR, but happy for you to push the other ones at any point.)

cachemeifyoucan replied:
I copy that exact line of code from original implementation from `CompilerInstance.cpp`. I don't know the exact problem but `OS` has to be moved here, otherwise PIPE just wont work on OutputBackend.

dexonsmith replied:
Looks to me like pipes do work:
```
% ./build/bin/clang -E -x c /dev/null | head    
# 1 "/dev/null"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 353 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "/dev/null" 2

% ./build/bin/clang -dM -E -x c /dev/null | head
#define _LP64 1
#define __APPLE_CC__ 6000
#define __APPLE__ 1
#define __ATOMIC_ACQUIRE 2
#define __ATOMIC_ACQ_REL 4
#define __ATOMIC_CONSUME 1
#define __ATOMIC_RELAXED 0
#define __ATOMIC_RELEASE 3
#define __ATOMIC_SEQ_CST 5
#define __BIGGEST_ALIGNMENT__ 16
```
Which specific tests are you looking at? I can take over fixing those ones.

cachemeifyoucan replied:
I was thinking just abandon the output backend parts since we can just merge in the upstream code which might be very different. For failing tests on this branch, you can look at:
`clang/test/Frontend/rewrite-includes-missing.c`

dexonsmith replied:
Okay, SGTM.

If it's just one test, you could also add `REQUIRES: disabled` with a FIXME saying that output backend breaks it. I seem to remember rewrite-includes being weird.

cachemeifyoucan committed:
[Frontend] Disable tests that are failing with current OutputBackend

Some tests are failing with current OutputBackend becasue the output
cannot be PIPE correctly. Disable the tests with a FIXME.
